[Site Isolation] Add basic support for loading webarchive

szewai opened the pull request:
#### db6a018c3ee4fd12cf56a3a1520889faace38993
<pre>
[Site Isolation] Add basic support for loading webarchive
<a href="https://bugs.webkit.org/show_bug.cgi?id=302978">https://bugs.webkit.org/show_bug.cgi?id=302978</a>
<a href="https://rdar.apple.com/165236186">rdar://165236186</a>

Reviewed by Ryosuke Niwa.

To make webarchive loading work with site isolation, make all frames (including cross-site ones) in a webarchive file
loaded within the same web process. When webarhive file is loaded in a web process, the process will be getting all
content of the file from network process, which includes subframe webarchive and resources. If UI process decides to
start loading of cross-site frame in another process, the new process won&apos;t have the resources and the load could fail.
To implement this, WebProcessPool::processForNavigation now returns source process for non-main-frame navigation when
page (main frame) is loading webarchive.

In current implementation, UI process would recreate FrameProcess when it notices frame&apos;s site does not match frame&apos;s
process&apos;s site in WebPageProxy::didCommitLoadForFrame. The problem is, for webarchive load, frame&apos;s site might not match
frame&apos;s process site as we don&apos;t swap process for iframe navigation. In this case when FrameProcess is recreated,
BrowsingContextGroup would try creating RemotePageProxy for the FrameProcess, so that there will be both RemotePageProxy
and WebPageProxy for the same page and same web process, which would lead to weird eror. To fix that, the patch add a
flag in FrameProcess to indicate the process is used for loading webarchive, and there is not need to recreate
FrameProcess for site mismatch in this case.

The patch also fixes an issue where UI process would try to swap in ephemeral webarchive data store for non-main-frame
navigation, see WebPageProxy::updateDataStoreForWebArchiveLoad.

API tests: SiteIsolation.LoadWebArchive
           SiteIsolation.LoadWebArchiveNestedFrame

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite):
(WebKit::BrowsingContextGroup::ensureProcessForSite):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/FrameProcess.cpp:
(WebKit::FrameProcess::FrameProcess):
* Source/WebKit/UIProcess/FrameProcess.h:
(WebKit::FrameProcess::isArchiveProcess const):
(WebKit::FrameProcess::create):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateDataStoreForWebArchiveLoad):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, LoadWebArchive)):
(TestWebKitAPI::(SiteIsolation, LoadWebArchiveNestedFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationLoadWebArchive.webarchive: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationLoadWebArchiveNestedFrame.webarchive: Added.

Canonical link: <a href="https://commits.webkit.org/303783@main">https://commits.webkit.org/303783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b46527f0a0555e651f2e447990cac70454a1867

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133567 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6072 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44726 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141127 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c3ee5aab-1bc3-41de-a30f-ae568893a059) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135437 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6593 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5936 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102183 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/19ee0f2f-dfbf-459d-b8e6-9dbe678b29d7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136514 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119751 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82979 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e9ddd3dc-772e-4390-a043-e4e70daadc98) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4562 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2162 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113698 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143776 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5744 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38444 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110561 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5824 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4929 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110739 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28083 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4414 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116002 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59493 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5796 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34317 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5643 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5887 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5751 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/FrameProcess.cpp:
```diff
@@ -35,18 +36,23 @@
+
+    if (!m_isArchiveProcess)
+        process.didStartUsingProcessForSiteIsolation(site, mainFrameSite);
     ASSERT(isSharedProcess() == process.isSharedProcess());
```
Looks like this assertion is missing when `!preferences.siteIsolationEnabled()`?

rniwa commented on Source/WebKit/UIProcess/FrameProcess.h:
```diff
@@ -47,20 +47,22 @@ class FrameProcess : public RefCountedAndCanMakeWeakPtr<FrameProcess> {
     friend class BrowsingContextGroup; // FrameProcess should not be created except by BrowsingContextGroup.
     static Ref<FrameProcess> create(WebProcessProxy& process, BrowsingContextGroup& group, const std::optional<WebCore::Site>& site, const WebCore::Site& mainFrameSite,
-        const WebPreferences& preferences, InjectBrowsingContextIntoProcess injectBrowsingContextIntoProcess)
+        const WebPreferences& preferences, LoadedWebArchive loadedWebArchive, InjectBrowsingContextIntoProcess injectBrowsingContextIntoProcess)
```
Don't we need to either forward-declare `LoadedWebArchive` or `#include` `LoadedWebArchive.h`?

rniwa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4968,18 +4968,66 @@ static std::optional<std::pair<Ref<API::WebsitePolicies>, Ref<WebProcessProxy>>>
+{
+    std::pair<RefPtr<WebsiteDataStore>, LoadedWebArchive> result { nullptr, LoadedWebArchive::No };
+    if (!protectedPreferences()->loadWebArchiveWithEphemeralStorageEnabled())
+        return result;
```
Tracking the value of `result` is rather difficult here.
I think we should add a helper function to return an explicit pair instead like:
```cpp
    auto makeResult = (RefPtr<WebsiteDataStore> dataStore, LoadedWebArchive loadedWebArchive)[] -> auto {
        return std::pair<RefPtr<WebsiteDataStore>, LoadedWebArchive> { dataStore, loadedWebArchive };
    };
    return makeResult(nullptr, LoadedWebArchive::No);
```

rniwa replied:
Or rather, should probably declare a `struct` with 3 members so that the code which uses this return value isn't cryptic:
```cpp
struct WebArchiveDataStoreResult {
    WebCore::ResourceError error;
    RefPtr<WebsiteDataStore> dataStore;
    LoadedWebArchive loadedWebArchive;
};
```

rniwa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4968,18 +4968,66 @@ static std::optional<std::pair<Ref<API::WebsitePolicies>, Ref<WebProcessProxy>>>
+    }
+
+#if PLATFORM(MAC)
+    // Client does not have access to the archive file.
```
Rather than adding a comment, define a descriptive local variable like this:
```cpp
bool clientDoesNotHaveAccessToArchiveFile = !isSubstituteDataWebArchive && isQuarantinedAndNotUserApproved(requestURL.fileSystemPath())
```

rniwa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5044,42 +5092,22 @@ void WebPageProxy::receivedNavigationActionPolicyDecision(WebProcessProxy& proce
+            receivedPolicyDecision(PolicyAction::Ignore, &navigation, std::nullopt, WTFMove(navigationAction), WillContinueLoadInNewProcess::No, std::nullopt, WTFMove(message), WTFMove(completionHandler));
+            return;
+        }
+        if (auto updatedWebsiteDataStore = result.value().first; updatedWebsiteDataStore && updatedWebsiteDataStore.get() != websiteDataStore.ptr()) {
```
This is hard to read. We should probably add a `struct` instead.

szewai committed:
[Site Isolation] Add basic support for loading webarchive
https://bugs.webkit.org/show_bug.cgi?id=302978
rdar://165236186

Reviewed by Ryosuke Niwa.

To make webarchive loading work with site isolation, make all frames (including cross-site ones) in a webarchive file
loaded within the same web process. When webarhive file is loaded in a web process, the process will be getting all
content of the file from network process, which includes subframe webarchive and resources. If UI process decides to
start loading of cross-site frame in another process, the new process won't have the resources and the load could fail.
To implement this, WebProcessPool::processForNavigation now returns source process for non-main-frame navigation when
page (main frame) is loading webarchive.

In current implementation, UI process would recreate FrameProcess when it notices frame's site does not match frame's
process's site in WebPageProxy::didCommitLoadForFrame. The problem is, for webarchive load, frame's site might not match
frame's process site as we don't swap process for iframe navigation. In this case when FrameProcess is recreated,
BrowsingContextGroup would try creating RemotePageProxy for the FrameProcess, so that there will be both RemotePageProxy
and WebPageProxy for the same page and same web process, which would lead to weird eror. To fix that, the patch add a
flag in FrameProcess to indicate the process is used for loading webarchive, and there is not need to recreate
FrameProcess for site mismatch in this case.

The patch also fixes an issue where UI process would try to swap in ephemeral webarchive data store for non-main-frame
navigation, see WebPageProxy::updateDataStoreForWebArchiveLoad.

API tests: SiteIsolation.LoadWebArchive
           SiteIsolation.LoadWebArchiveNestedFrame

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite):
(WebKit::BrowsingContextGroup::ensureProcessForSite):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/FrameProcess.cpp:
(WebKit::FrameProcess::FrameProcess):
* Source/WebKit/UIProcess/FrameProcess.h:
(WebKit::FrameProcess::isArchiveProcess const):
(WebKit::FrameProcess::create):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateDataStoreForWebArchiveLoad):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, LoadWebArchive)):
(TestWebKitAPI::(SiteIsolation, LoadWebArchiveNestedFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationLoadWebArchive.webarchive: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationLoadWebArchiveNestedFrame.webarchive: Added.

Canonical link: https://commits.webkit.org/303783@main
